rfcs/text/0007-multiple-users.md

```diff
@@ -0,0 +1,769 @@
+
+Any event that has user(s) in it should always populate the array field `related.user`
+with all usernames seen on the event. Note that this field is not a nesting of
+all user fields, it's a flat array meant to contain user identifiers.
+
+Taking the example from `user.changes` again, and populating `related.user` as well,
+the event now looks like:
+
+```JSON
+{
+  "user": {
+    "name": "alice",
+    "id": "1001",
+    "effective": {
+      "name": "root",
+      "id": "1"
+    },
+    "target": {
+      "name": "bob",
+      "id": "1002"
+    },
+    "changes": {
+      "name": "bob.barker"
+    }
+  },
+  "related": { "user": ["alice", "root", "bob", "bob.barker"] }
+}
+```
+
+## Source data
+
+Here are some concrete examples of events with multiple users and user roles.
+Note that the design of these fields is meant to allow all of their use at the
+same time, *when needed*. However if events don't contain all user roles because
+they're spread out across events, only the fields relevant to each event should be used.
+
+### Linux IAM and privilege escalation
+
+Here's a typical set of log about a user creation on Linux.
+
+```
+Sep 29 19:55:09 localhost sudo: vagrant : TTY=pts/0 ; PWD=/home/vagrant ; USER=root ; COMMAND=/sbin/useradd test-user -p test-password
+Sep 29 19:55:09 localhost sudo: pam_unix(sudo:session): session opened for user root by vagrant(uid=0)
+Sep 29 19:55:09 localhost useradd[2097]: new group: name=test-user, GID=1001
+Sep 29 19:55:09 localhost useradd[2097]: new user: name=test-user, UID=1001, GID=1001, home=/home/test-user, shell=/bin/bash
+Sep 29 19:55:09 localhost sudo: pam_unix(sudo:session): session closed for user root
+```
+
+#### Logical events
+
+A solution that coalesces log events to produce higher level logical events could
+capture them all in the following way.
+
+Group creation:
+
+```JSON
+{
+  "event": {
+    "kind": "event",
+    "category": ["iam"],
+    "type": ["group", "creation"]
+  },
+  "group": {
+    "name": "test-user",
+    "id": "1001"
+  },
+  "user": {
+    "name": "vagrant",
+    "id": "1000",
+    "effective": {
+      "name": "root",
+      "id": "0"
+    },
+  },
+  "related": { "user": ["vagrant", "root"] }
+}
+```
+
+User creation:
+
+```JSON
+{
+  "event": {
+    "kind": "event",
+    "category": ["iam"],
+    "type": ["user", "creation"]
+  },
+  "user": {
+    "name": "vagrant",
+    "id": "1000",
+    "effective": {
+      "name": "root",
+      "id": "0"
+    },
+    "target": {
+      "name": "test-user",
+      "id": "1001",
+      "group": {
+        "name": "test-user",
+        "id": "1001"
+      }
+    }
+  },
+  "related": { "user": ["vagrant", "root", "test-user"] }
+}
+```
+
+#### Raw events
+
+A solution that produces one event per log without coalescing would instead only
+represent users with the information available in the given log event.
+
+event 1:
+
+```JSON
+{
+  "event": {
+    "kind": "event",
+    "category": ["process"],
+    "event": "creation"
```

Comment:
```suggestion
    "type": ["creation"]
```

Reply:
I will not drop a 🤦  on each of these, but the temptation is high 😂

Thanks for catching that.